concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+    BlockHash
+successorBlockHash bhRound bhEpoch bhParent = computeBlockHash bhh
+  where
+    bhh = getHash BlockHeader{..}
+
+-- |A finalization entry that witnesses that a block has been finalized with quorum certificates
+-- for two consecutive rounds. The finalization entry includes a proof that the blocks are in
+-- consecutive rounds so that the entry can be validated without the second block.
+--
+-- The following invariants hold:
+--
+-- - @qcRound feSuccessorQuorumCertificate == qcRound feFinalizedQuorumCertificate + 1@
+-- - @qcBlock feSuccessorQuorumCertificate == successorBlockHash (qcRound feSuccessorQuorumCertificate) (qcEpoch feSuccessorQuorumCertificate) (qcBlock feFinalizedQuorumCertificate) feSuccessorProof@
+data FinalizationEntry = FinalizationEntry
+    { -- |Quorum certificate for the finalized block.
+      feFinalizedQuorumCertificate :: !QuorumCertificate,
+      -- |Quorum certificate for the successor block.
+      feSuccessorQuorumCertificate :: !QuorumCertificate,
+      -- |Proof that establishes the successor block is the immediate successor of the finalized
+      -- block (without further knowledge of the successor block beyond its hash).
+      feSuccessorProof :: !SuccessorProof
+    }
+    deriving (Eq, Show)
+
+instance Serialize FinalizationEntry where
+    put FinalizationEntry{..} = do
+        put feFinalizedQuorumCertificate
+        let QuorumCertificate{..} = feSuccessorQuorumCertificate
+        put qcEpoch
+        put qcAggregateSignature
+        put qcSignatories
+        put feSuccessorProof
+    get = do
+        feFinalizedQuorumCertificate <- get
+        qcEpoch <- get
+        qcAggregateSignature <- get
+        qcSignatories <- get
+        feSuccessorProof <- get
+        let sqcRound = qcRound feFinalizedQuorumCertificate + 1
+        let feSuccessorQuorumCertificate =
+                QuorumCertificate
+                    { qcRound = sqcRound,
+                      qcBlock =
+                        successorBlockHash
+                            sqcRound
+                            qcEpoch
+                            (qcBlock feFinalizedQuorumCertificate)
+                            feSuccessorProof,
+                      ..
+                    }
+        return FinalizationEntry{..}
+
+instance HashableTo Hash.Hash (Optionally FinalizationEntry) where
+    getHash Absent = Hash.hash $ encode (0 :: Word8)
+    getHash (Present fe) = Hash.hash $ runPut $ do
+        putWord8 1
+        put fe
+
+data TimeoutSignatureMessage = TimeoutSignatureMessage
+    { -- |Hash of the genesisBlock
+      tsmGenesis :: !BlockHash,
+      -- |Round number of the timed-out round
+      tsmRound :: !Round,
+      -- |Round number of the highest known valid quorum certificate
+      tsmQCRound :: !Round
+    }
+    deriving (Eq, Show)
+
+-- |Compute the byte representation of a 'TimeoutSignatureMessage' that is actually signed.
+-- TODO: check that this cannot collide with other signed messages.
```

Comment:
When do you expect to address these TODOs?

Please put them in an issue if not now. They will be forgotten in the code here.

Reply:
Todo removed. Will add to an issue.